RecoMET/METPUSubtraction/src/PFMETAlgorithmMVA.cc

```diff
@@ -139,255 +165,115 @@ void PFMETAlgorithmMVA::setInput(const std::vector<reco::PUSubMETCandInfo>& lept
-				 double puSumEt, double puMEt, double puPhi,
-				 double pucSumEt, double pucU, double pucPhi,
-				 double jet1Pt, double jet1Eta, double jet1Phi,
-				 double jet2Pt, double jet2Eta, double jet2Phi,
-				 double numJetsPtGt30, double numJets, 
-				 double numVertices)
+//-------------------------------------------------------------------------------
+float* PFMETAlgorithmMVA::createFloatVector(std::vector<std::string> variableNames)
 {
-  // protection against "empty events"
-  if ( pfSumEt < 1. ) pfSumEt = 1.;
-  
-  pfSumEt_       = pfSumEt;
-  pfU_           = pfU;
-  pfPhi_         = pfPhi;
-  tkSumEt_       = tkSumEt/pfSumEt_;
-  tkU_           = tkU;
-  tkPhi_         = tkPhi;
-  npuSumEt_      = npuSumEt/pfSumEt_;
-  npuU_          = npuU;
-  npuPhi_        = npuPhi;
-  puSumEt_       = puSumEt/pfSumEt_;
-  puMEt_         = puMEt;
-  puPhi_         = puPhi;
-  pucSumEt_      = pucSumEt/pfSumEt_;
-  pucU_          = pucU;
-  pucPhi_        = pucPhi;
-  jet1Pt_        = jet1Pt;
-  jet1Eta_       = jet1Eta;
-  jet1Phi_       = jet1Phi;
-  jet2Pt_        = jet2Pt;
-  jet2Eta_       = jet2Eta;
-  jet2Phi_       = jet2Phi;
-  numJetsPtGt30_ = numJetsPtGt30;
-  numJets_       = numJets;
-  numVertices_   = numVertices;
+    float* floatVector = new float[variableNames.size()];
+    int i = 0;
+    for(auto variableName: variableNames)
+    {
+        floatVector[i++] = var_[variableName];
+    }
+    return floatVector;
 }
-//-------------------------------------------------------------------------------
 
 //-------------------------------------------------------------------------------
 void PFMETAlgorithmMVA::evaluateMVA()
 {
   // CV: MVAs needs to be evaluated in order { DPhi, U1, CovU1, CovU2 }
   //     as MVA for U1 (CovU1, CovU2) uses output of DPhi (DPhi and U1) MVA
-  evaluateDPhi();
-  evaluateU();
-  evaluateCovU1();
-  evaluateCovU2();
+  mvaOutputDPhi_  = GetResponse(mvaReaderDPhi_, varForDPhi_);
+  var_["PhiCor_UPhi"] = var_["particleFlow_UPhi"] + mvaOutputDPhi_;
+  mvaOutputU_     = GetResponse(mvaReaderU_, varForU_);
+  var_["RecoilCor_U"] = var_["particleFlow_U"] * mvaOutputU_;
+  var_["RecoilCor_UPhi"] = var_["PhiCor_UPhi"];
+  mvaOutputCovU1_ = GetResponse(mvaReaderCovU1_, varForCovU1_)* mvaOutputU_ * var_["particleFlow_U"];
+  mvaOutputCovU2_ = GetResponse(mvaReaderCovU2_, varForCovU2_)* mvaOutputU_ * var_["particleFlow_U"];
+
 
   // compute MET(Photon check)
   if(hasPhotons_) { 
     //Fix events with unphysical properties
     double sumLeptonPt = std::max(sqrt(sumLeptonPx_*sumLeptonPx_+sumLeptonPy_*sumLeptonPy_),1.);
-    if(tkU_/sumLeptonPt < 0.1 || npuU_/sumLeptonPt <  0.1 ) mvaOutputU_      = 1.;
-    if(tkU_/sumLeptonPt < 0.1 || npuU_/sumLeptonPt <  0.1 ) mvaOutputDPhi_   = 0.;
+    if(var_["track_U"]/sumLeptonPt < 0.1 || var_["noPileUp_U"]/sumLeptonPt <  0.1 ) mvaOutputU_      = 1.;
+    if(var_["track_U"]/sumLeptonPt < 0.1 || var_["noPileUp_U"]/sumLeptonPt <  0.1 ) mvaOutputDPhi_   = 0.;
```

Comment:
@slava77 do you have any concern about the protection currently used for very rare photon cases?